plugins/RandomTriggerCandidateMaker.cpp

```diff
@@ -0,0 +1,171 @@
+  register_command("start", &RandomTriggerCandidateMaker::do_start);
+  register_command("stop", &RandomTriggerCandidateMaker::do_stop);
+  register_command("scrap", &RandomTriggerCandidateMaker::do_scrap);
+}
+
+void
+RandomTriggerCandidateMaker::init(const nlohmann::json& obj)
+{
+  m_time_sync_source.reset(new appfwk::DAQSource<dfmessages::TimeSync>(appfwk::queue_inst(obj,"time_sync_source")));
+  m_trigger_candidate_sink.reset(new appfwk::DAQSink<triggeralgs::TriggerCandidate>(appfwk::queue_inst(obj, "trigger_candidate_sink")));
+}
+
+void
+RandomTriggerCandidateMaker::get_info(opmonlib::InfoCollector& /*ci*/, int /*level*/)
+{
+}
+
+void
+RandomTriggerCandidateMaker::do_configure(const nlohmann::json& obj)
+{
+  m_conf = obj.get<randomtriggercandidatemaker::ConfParams>();
+}
+
+void
+RandomTriggerCandidateMaker::do_start(const nlohmann::json& obj)
+{
+  m_run_number = obj.value<dunedaq::dataformats::run_number_t>("run", 0);
+
+  m_running_flag.store(true);
+
+  switch (m_conf.timestamp_method) {
+    case randomtriggercandidatemaker::timestamp_estimation::kTimeSync:
+      TLOG_DEBUG(0) << "Creating TimestampEstimator";
+      m_timestamp_estimator.reset(new TimestampEstimator(m_time_sync_source, m_conf.clock_frequency_hz));
+      break;
+    case randomtriggercandidatemaker::timestamp_estimation::kSystemClock:
+      TLOG_DEBUG(0) << "Creating TimestampEstimatorSystem";
+      m_timestamp_estimator.reset(new TimestampEstimatorSystem(m_conf.clock_frequency_hz));
+      break;
+  }
+
+  m_send_trigger_candidates_thread = std::thread(&RandomTriggerCandidateMaker::send_trigger_candidates, this);
+  pthread_setname_np(m_send_trigger_candidates_thread.native_handle(), "random-tc-maker");
+}
+
+void
+RandomTriggerCandidateMaker::do_stop(const nlohmann::json& /*obj*/)
+{
+  m_running_flag.store(false);
+
+  m_send_trigger_candidates_thread.join();
+
+  m_timestamp_estimator.reset(nullptr); // Calls TimestampEstimator dtor
+}
+
+void
+RandomTriggerCandidateMaker::do_scrap(const nlohmann::json& /*obj*/)
+{
+  m_configured_flag.store(false);
+}
+
+triggeralgs::TriggerCandidate
+RandomTriggerCandidateMaker::create_candidate(dfmessages::timestamp_t timestamp)
+{
+  triggeralgs::TriggerCandidate candidate;
+  candidate.time_start = timestamp;
+  candidate.time_end = timestamp;
+  candidate.time_candidate = timestamp;
+  candidate.detid = { 0 };
+  candidate.type = triggeralgs::TriggerCandidateType::kRandom;
+  candidate.algorithm = 0;
+  candidate.version = 0;
+
+  return candidate;
+}
+
+int
+RandomTriggerCandidateMaker::get_interval(std::mt19937 &gen)
+{
+  switch (m_conf.time_distribution) {
+    case randomtriggercandidatemaker::distribution_type::kUniform:
+      return m_conf.trigger_interval_ticks;
+    case randomtriggercandidatemaker::distribution_type::kPoisson:
+      std::exponential_distribution<double> d(1.0/m_conf.trigger_interval_ticks);
+      return (int)(0.5+d(gen));
+  }
+}
```

Comment:
gcc gives me a warning here:

```
/home/rodrigues/dune/daq/dunedaq-v2.4.0-testrels/trigger/sourcecode/trigger/plugins/RandomTriggerCandidateMaker.cpp: In member function ‘int dunedaq::trigger::RandomTriggerCandidateMaker::get_interval(std::mt19937&)’:
/home/rodrigues/dune/daq/dunedaq-v2.4.0-testrels/trigger/sourcecode/trigger/plugins/RandomTriggerCandidateMaker.cpp:132:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
```

Could you silence it (with a `default` or something else)?

Reply:
Curious, GCC should know all elements of the enum are implemented. I'll have default treated as kUniform with message printed to the log.